crates/oxc_query/src/vertex.rs

```diff
@@ -0,0 +1,319 @@
+            }
+        }
+    }
+
+    pub fn make_url(attr: &'a JSXAttribute<'a>) -> Option<Self> {
+        let Some(maybe_url) = jsx_attribute_to_constant_string(attr) else { return None };
+        let Ok(parsed_url) = Url::parse(&maybe_url) else { return None };
+        return Some(Vertex::URL(Rc::new(parsed_url)));
+    }
+
+    pub fn as_constant_string(&self) -> Option<String> {
+        match &self {
+            Vertex::Expression(expr) => expr_to_maybe_const_string(expr),
+            _ => None,
+        }
+    }
+}
+
+impl trustfall::provider::Typename for Vertex<'_> {
+    fn typename(&self) -> &'static str {
+        self.make_type_name()
+    }
+}
+
+impl<'a> From<AstNode<'a>> for Vertex<'a> {
+    fn from(ast_node: AstNode<'a>) -> Self {
+        match ast_node.kind() {
+            AstKind::ReturnStatement(return_statement) => Self::ReturnStatementAST(
+                ReturnStatementVertex { ast_node: Some(ast_node), return_statement }.into(),
+            ),
+            AstKind::JSXElement(element) => {
+                Self::JSXElement(JSXElementVertex { ast_node: Some(ast_node), element }.into())
+            }
+            AstKind::TSInterfaceDeclaration(interface) => {
+                Self::Interface(InterfaceVertex { ast_node: Some(ast_node), interface }.into())
+            }
+            AstKind::TSTypeAnnotation(type_annotation) => Self::TypeAnnotation(
+                TypeAnnotationVertex { ast_node: Some(ast_node), type_annotation }.into(),
+            ),
+            AstKind::VariableDeclarator(variable_declaration) => Self::VariableDeclaration(
+                VariableDeclarationVertex { ast_node: Some(ast_node), variable_declaration }.into(),
+            ),
+            AstKind::ModuleDeclaration(md)
+                if matches!(md, ModuleDeclaration::ImportDeclaration(_)) =>
+            {
+                let ModuleDeclaration::ImportDeclaration(import) = md else { unreachable!() };
+                Self::Import(ImportVertex { ast_node: Some(ast_node), import }.into())
+            }
+            AstKind::Class(class) => {
+                Self::Class(ClassVertex { ast_node: Some(ast_node), class }.into())
+            }
+            _ => Vertex::ASTNode(ast_node),
+        }
+    }
+}
+
+impl<'a> From<&'a Expression<'a>> for Vertex<'a> {
+    fn from(expr: &'a Expression<'a>) -> Self {
+        // FIXME: We just get rid of all parentheses here, but we shouldn't do that...
+
+        // NOTE: When string literal / template lietal is added, add to as_constant_string
+        match &expr.get_inner_expression() {
+            Expression::ObjectExpression(objexpr) => Vertex::ObjectLiteral(objexpr),
+            Expression::JSXElement(element) => {
+                Vertex::JSXElement(JSXElementVertex { ast_node: None, element }.into())
+            }
+            _ => Vertex::Expression(expr),
+        }
+    }
+}
+
+pub trait TypeName {
+    fn type_name(&self) -> &'static str;
+}
```

Comment:
Why define a custom trait for this instead of using the identical one from Trustfall?

Reply:
fixed in [88ba9f7](https://github.com/web-infra-dev/oxc/pull/599/commits/88ba9f742679dcf2b55dedbf9d41cffb5f200cfb)